File: implement-shell-tools/ls/ls.js
import { program } from "commander";
import { promises as fs } from "node:fs";

program
    .name("ls")
    .description("List all files name")
    .option("-1", "A new line for each file")
    .option("-a", "Show hidden files")
    .argument("[path]", "The directory to list");

program.parse();

let path = "";
program.args.length == 0 ? path = '.' : path = program.args[0];
const flags = program.opts()

try {
    let files = await fs.readdir(path);
    if (!flags["a"]) {
        files = files.filter(file => !file.startsWith("."))
    }
    if (flags["1"]) {
        console.log(files.join("\n"));
    } else {
        console.log(files.join("       "));
    }
} catch (err) {
    console.error(`ls - ${path}: ${err.message}`)
}

File: implement-shell-tools/ls/ls.py
import sys
import argparse
import os

parser = argparse.ArgumentParser(
    prog="ls",
    description="List all files name",
)

parser.add_argument("-1", dest="one", action="store_true", help="A new line for each file")
parser.add_argument("-a", action="store_true", help="Show hidden files")
parser.add_argument("path", nargs="?", default=".", help="The directory to list")

args = parser.parse_args()
path = args.path

try:
    files = sorted(os.listdir(path))
    if not args.a:
        files = [file for file in files if not file.startswith(".")]
    if args.one:
        print("\n".join(files))
    else:
        print("       ".join(files))

except Exception as e:
    print(f"Error reading {path}: {e}")